pkg/util/voljournal.go

```diff
@@ -0,0 +1,394 @@
+	imageUUID := strings.TrimPrefix(volName, cj.namingPrefix)
+	err := RemoveObject(monitors, id, key, pool, cj.cephUUIDDirectoryPrefix+imageUUID)
+	if err != nil {
+		if _, ok := err.(ErrObjectNotFound); !ok {
+			klog.Errorf("failed removing oMap %s (%s)", cj.cephUUIDDirectoryPrefix+imageUUID, err)
+			return err
+		}
+	}
+
+	// delete the request name key (last, inverse of create order)
+	err = RemoveOMapKey(monitors, id, key, pool, cj.csiDirectory,
+		cj.csiNameKeyPrefix+reqName)
+	if err != nil {
+		klog.Errorf("failed removing oMap key %s (%s)", cj.csiNameKeyPrefix+reqName, err)
+		return err
+	}
+
+	return err
+}
+
+// reserveOMapName creates an omap with passed in oMapNamePrefix and a generated <uuid>.
+// It ensures generated omap name does not already exist and if conflicts are detected, a set
+// number of retires with newer uuids are attempted before returning an error
+func reserveOMapName(monitors, id, key, pool, oMapNamePrefix string) (string, error) {
+	var iterUUID string
+
+	maxAttempts := 5
+	attempt := 1
+	for attempt <= maxAttempts {
+		// generate a uuid for the image name
+		iterUUID = uuid.NewUUID().String()
+
+		err := CreateObject(monitors, id, key, pool, oMapNamePrefix+iterUUID)
+		if err != nil {
+			if _, ok := err.(ErrObjectExists); ok {
+				attempt++
+				// try again with a different uuid, for maxAttempts tries
+				klog.V(4).Infof("uuid (%s) conflict detected, retrying (attempt %d of %d)",
+					iterUUID, attempt, maxAttempts)
+				continue
+			}
+
+			return "", err
+		}
+
+		break
+	}
+
+	if attempt > maxAttempts {
+		return "", errors.New("uuid conflicts exceeds retry threshold")
+	}
+
+	return iterUUID, nil
+}
+
+/*
+ReserveName adds respective entries to the csiDirectory omaps, post generating a target
+UUIDDirectory for use. Further, these functions update the UUIDDirectory omaps, to store back
+pointers to the CSI generated request names.
+
+NOTE: As the function manipulates omaps, it should be called with a lock against the request name
+held, to prevent parallel operations from modifying the state of the omaps for this request name.
+
+Return values:
+	- string: Contains the UUID that was reserved for the passed in reqName
+	- error: non-nil in case of any errors
+*/
+func (cj *CSIJournal) ReserveName(monitors, id, key, pool, reqName, parentName string) (string, error) {
+	var snapSource bool
+
+	if parentName != "" {
+		if cj.cephSnapSourceKey == "" {
+			err := errors.New("invalid snapshot request on volume journal")
+			return "", err
+		}
+		snapSource = true
+	}
+
+	// Create the UUID based omap first, to reserve the same and avoid conflicts
+	// NOTE: If any service loss occurs post creation of the UUID directory, and before
+	// setting the request name key (csiNameKey) to point back to the UUID directory, the
+	// UUID directory key will be leaked
+	volUUID, err := reserveOMapName(monitors, id, key, pool, cj.cephUUIDDirectoryPrefix)
```

Comment:
Before reserving or setting omap, dont we need to check whether it exist in the omap. Isnt it ?

Reply:
Yes that is correct. If used unchecked then existing omaps **will** get overwritten.

Hence the usage pattern in the code is as follows,
- Lock on the [requestName](https://github.com/ceph/ceph-csi/blob/2d560ba0872569a46f0d890491c8aee702f2e5b6/pkg/rbd/controllerserver.go#L113) to prevent parallel requests for the same name (as mentioned in the [comment](https://github.com/ceph/ceph-csi/pull/381/files#diff-dbad340382ae55fa4ae6cc5429ca80dbR296))
- [checkVolExists](https://github.com/ceph/ceph-csi/blob/2d560ba0872569a46f0d890491c8aee702f2e5b6/pkg/rbd/controllerserver.go#L125), which calls [CheckReservation](https://github.com/ceph/ceph-csi/pull/381/files#diff-a6cef7e66fbdf22f7065bf167606d504R182)
- On failure to find a reservation, call [reserveVol](https://github.com/ceph/ceph-csi/blob/2d560ba0872569a46f0d890491c8aee702f2e5b6/pkg/rbd/controllerserver.go#L143) which in turn calls ReserveName

Hence, the usage ensures that reservation is checked and on failure to find one, ReserveName is called.